README.md
[![build status](https://secure.travis-ci.org/TooTallNate/node-cgi.png)](http://travis-ci.org/TooTallNate/node-cgi)
node-cgi
========
### An http/stack/connect layer to invoke and serve CGI executables.


This module implements [RFC 3875][rfc3875], and offers an easy interface to run
and serve CGI executables using [Node][]'s HTTP server. I wrote this so I could
directly serve [GitWeb][node-gitweb] through Node.


CGI Scripts?
------------

If you're not familiar with CGI scripts, they're simply executable files that
get invoked by a web server with client requests. The script has Environment
Variables set that indicate information about the HTTP request the client has sent.

Here's what a simple "Hello World" CGI script in `sh` would look like:

``` bash
#!/bin/sh

# Headers are written first. The special "Status"
# header indicates the response status code
echo "Status: 200"
echo "Content-Type: text/plain"
echo
    
# Followed by a response body
echo "Hello World!"
```

Let's call it `hello.cgi`. Be sure to make it executable with `chmod +x hello.cgi`!


Invoking "The Script" with Node
-------------------------------

Now, we need to set up our Node HTTP server. For every request sent to the server,
our `hello.cgi` script will be invoked, and the response will be sent back to the
HTTP client:

``` javascript
var http = require('http');
var cgi = require('cgi');

http.createServer(
  cgi('hello.cgi')
).listen(80);
```

This will set up a CGI handler with the default options.


License
-------

(The MIT License)

Copyright (c) 2012 Nathan Rajlich &lt;nathan@tootallnate.net&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[Node]: http://nodejs.org
[node-gitweb]: https://github.com/TooTallNate/node-gitweb
[rfc3875]: http://tools.ietf.org/html/rfc3875


History.md

0.3.1 / 2015-01-29
==================

  * cgi: ensure that the spawn opts don't trump the cgi options object
  * cgi: use .pipe() for stderr
  * cgi: use `%o` formatter for debug() calls
  * package: update "extend" to v2.0.0
  * package: use ".js" in "main" field
  * package: allow any "debug" v2

0.3.0 / 2014-05-22
==================

  * test: fix "env" test
  * cgi: pass the `options` object directly to spawn()
  * cgi: use "extend" module instead of inlined code
  * cgi: ensure a leading / char for `PATH_INFO` var
  * add LICENSE file
  * package: add "license" field
  * package: remove "engines" field
  * package: use ~ instead of >= for dependencies
  * package: remove period from "description"
  * Added `REMOTE_ADDR` and `REMOTE_PORT` to env variables


test/env.js
var http = require('http');
var cgi  = require('../cgi');
var connect = require('connect');

var PORT = 5555;
var app = connect()
  .use(connect.logger())
  .use(cgi(__dirname + '/cgi-bin/printenv.cgi'));

var server = http.createServer(app);

server.listen(PORT, function() {
  console.log('server listening');

  var req = http.get({
    port: PORT,
    path: '/?test=1'
  });

  req.on('response', function (res) {
    console.log(res.headers);
    res.pipe(process.stdout);
    res.on('end', server.close.bind(server));
  });

  req.end();
});


test/proxy.js
var fs = require('fs');
var http = require('http');
var https = require('https');
var cgi = require('cgi');

// The HTTPS SSL options
var options = {
  key: fs.readFileSync(__dirname + '/ssl.key'),
  cert: fs.readFileSync(__dirname + '/ssl.crt')
}

var hander = cgi(__dirname + '/cgi-bin/nph-proxy.cgi', {
  nph: true,
  env: {
    //RUNNING_ON_SSL_SERVER: 1
  }
});

http.createServer(handler).listen(80);
https.createServer(options, handler).listen(443);


bufferHelpers.js
/**
 * Accepts an arbitrary number of Buffer instances, and returns a single
 * Buffer instance with the contents of all the passed Buffers concatenated
 * into it.
 */
function bufferConcat() {
  var size = 0;
  var l=arguments.length;
  for (var i=0; i<l; i++) {
    size += arguments[i].length;
  }
  var rtn = new Buffer(size);
  var pos = 0;
  for (var i=0; i<l; i++) {
    arguments[i].copy(rtn, pos, 0);
    pos += arguments[i].length;
  }
  return rtn;
}
exports.bufferConcat = bufferConcat;

/**
 * A naiive 'Buffer.indexOf' function. Requires both the needle and haystack
 * to be Buffer instances.
 */
function bufferIndexOf(haystack, needle) {
  for (var i=0, l=haystack.length-needle.length+1; i<l; i++) {
    var good = true;
    for (var j=0, n=needle.length; j<n; j++) {
      if (haystack[i+j] !== needle[j]) {
        good = false;
        break;
      }
    }
    if (good) return i;
  }
  return -1;
}
exports.bufferIndexOf = bufferIndexOf;

cgi.js

/**
 * Module dependencies.
 */

var url = require('url');
var extend = require('extend');
var debug = require('debug')('cgi');
var spawn = require('child_process').spawn;
var CGIParser = require('./parser');

/**
 * Module exports.
 */

module.exports = cgi;

/**
 * Constants.
 */

var SERVER_SOFTWARE = 'Node/' + process.version;
var SERVER_PROTOCOL = 'HTTP/1.1';
var GATEWAY_INTERFACE = 'CGI/1.1';

function cgi(cgiBin, options) {
  options = extend({}, cgi.DEFAULTS, options);

  return function layer(req, res, next) {
    if (!next) {
      // define a default "next" handler if none was passed
      next = function(err) {
        debug('"next" called: %o', err);
        res.writeHead(404, { "Content-Type": "text/plain" });
        res.end("Not Found\n");
      };
    }
    if (!req.hasOwnProperty("uri")) { req.uri = url.parse(req.url); }
    if (req.uri.pathname.substring(0, options.mountPoint.length) !== options.mountPoint) return next();
    debug('handling HTTP request: %o', req.url);

    var host = (req.headers.host || '').split(':');
    var address = host[0];
    var port = host[1];
    if ((!address || !port) && typeof this.address == 'function') {
      var serverAddress = this.address();
      debug('server address and port: %o', serverAddress);
      if (!address) address = serverAddress.address;
      if (!port) port = serverAddress.port;
    }

    // Take environment variables from the current server process
    var env = extend({}, process.env);

    // Determine the correct PATH_INFO variable.
    // It must be prepended with a "/" char as per:
    //   https://tools.ietf.org/html/rfc3875#section-4.1.5
    var pathInfo = req.uri.pathname.substring(options.mountPoint.length);
    if ('/' !== pathInfo[0]) pathInfo = '/' + pathInfo;
    debug('calculated PATH_INFO variable: %o', pathInfo);

    // These meta-variables below can be overwritten by a
    // user's 'env' object in options
    extend(env, {
      GATEWAY_INTERFACE:  GATEWAY_INTERFACE,
      SCRIPT_NAME:        options.mountPoint,
      PATH_INFO:          pathInfo,
      SERVER_NAME:        address || 'unknown',
      SERVER_PORT:        port || 80,
      SERVER_PROTOCOL:    SERVER_PROTOCOL,
      SERVER_SOFTWARE:    SERVER_SOFTWARE
    });

    // The client HTTP request headers are attached to the env as well,
    // in the format: "User-Agent" -> "HTTP_USER_AGENT"
    for (var header in req.headers) {
      var name = 'HTTP_' + header.toUpperCase().replace(/-/g, '_');
      env[name] = req.headers[header];
    }

    // Now add the user-specified env variables
    if (options.env) extend(env, options.env);

    // These final environment variables take precedence over user-specified ones.
    env.REQUEST_METHOD = req.method;
    env.QUERY_STRING = req.uri.query || '';
    env.REMOTE_ADDR = req.connection.remoteAddress;
    env.REMOTE_PORT = req.connection.remotePort;
    if ('content-length' in req.headers) {
      env.CONTENT_LENGTH = req.headers['content-length'];
    }
    if ('content-type' in req.headers) {
      env.CONTENT_TYPE = req.headers['content-type'];
    }
    if ('authorization' in req.headers) {
      var auth = req.headers.authorization.split(' ');
      env.AUTH_TYPE = auth[0];
      //var unbase = new Buffer(auth[1], 'base64').toString().split(':');
    }

    var opts = extend({}, options);

    // Now we can spawn the CGI executable
    debug('env: %o', env);
    opts.env = env;

    var cgiSpawn = spawn(cgiBin, opts.args, opts);
    debug('cgi spawn (pid: %o)', cgiSpawn.pid);

    // The request body is piped to 'stdin' of the CGI spawn
    req.pipe(cgiSpawn.stdin);

    // If `options.stderr` is set to a Stream instance, then re-emit the
    // 'data' events onto the stream.
    if (options.stderr) {
      cgiSpawn.stderr.pipe(options.stderr);
    }

    // A proper CGI script is supposed to print headers to 'stdout'
    // followed by a blank line, then a response body.
    var cgiResult;
    if (!options.nph) {
      cgiResult = new CGIParser(cgiSpawn.stdout);

      // When the blank line after the headers has been parsed, then
      // the 'headers' event is emitted with a Headers instance.
      cgiResult.on('headers', function(headers) {
        headers.forEach(function(header) {
          // Don't set the 'Status' header. It's special, and should be
          // used to set the HTTP response code below.
          if (header.key === 'Status') return;
          res.setHeader(header.key, header.value);
        });

        // set the response status code
        res.statusCode = parseInt(headers.status, 10) || 200;

        // The response body is piped to the response body of the HTTP request
        cgiResult.pipe(res);
      });
    } else {
      // If it's an NPH script, then responsibility of the HTTP response is
      // completely passed off to the child process.
      cgiSpawn.stdout.pipe(res.connection);
    }

    cgiSpawn.on('exit', function(code, signal) {
      debug('cgi spawn %o "exit" event (code %o) (signal %o)', cgiSpawn.pid, code, signal);
      // TODO: react on a failure status code (dump stderr to the response?)
    });

    cgiSpawn.stdout.on('end', function () {
      // clean up event listeners upon the "end" event
      debug('cgi spawn %o stdout "end" event', cgiSpawn.pid);
      if (cgiResult) {
        cgiResult.cleanup();
      }
      //if (options.stderr) {
      //  cgiSpawn.stderr.unpipe(options.stderr);
      //}
    });
  };
}

// The default config options to use for each `cgi()` call.
cgi.DEFAULTS = {
  // The 'cgi' handler will take effect when the req.url begins with "mountPoint"
  mountPoint: '/',
  // Any additional variables to insert into the CGI script's Environment
  env: {},
  // Set to 'true' if the CGI script is an NPH script
  nph: false,
  // Set to a `Stream` instance if you want to log stderr of the CGI script somewhere
  stderr: null,
  // A list of arguments for the cgi bin to be used by spawn
  args: []
};
